### PR TITLE
ART-188: Added support for cPanel & WHM Api Headers

### PR DESCRIPTION
**Pull request by tomgreen98**
Also:

Added simpler way to convert the header to an object for
use with some frameworks.
Updated and fixed the examples in the readme.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,4 +40,10 @@ Correctly exported the utils modules.
 
 # v3.0.9
 
-Updated the library version to remove low security threat
\ No newline at end of file
+Updated the library version to remove low security threat
+
+# v3.0.10
+
+Add support for cPanel and Whm Api tokens.
```
Comment (laurencesimoncpanel) [inline CHANGELOG.md]: Whm -> WHM
Api -> API

File: README.md [hunk 2]
```diff
@@ -16,17 +16,18 @@ yarn add @cpanel/api
...
-
+import {
+    Argument,
+    WhmApiResponse,
+    WhmApiRequest,
+    WhmApiType,
+    WhmApiTokenHeader
+} from "@cpanel/api";
+
+const token = "...paste your api token here...";
```
Comment (laurencesimoncpanel) [inline README.md]: api -> API

File: README.md [hunk 3]
```diff
@@ -61,17 +61,18 @@ fetch('http://my-cpanel-server.com:2087', {
...
-
+let {
+    Argument,
+    WhmApiResponse,
+    WhmApiRequest,
+    WhmApiType,
+    WhmApiTokenHeader
+} = require("@cpanel/api");
+
+const token = "...paste your api token here...";
```
Comment (laurencesimoncpanel) [inline README.md]: api -> API

File: README.md [hunk 4]
```diff
@@ -61,17 +61,18 @@ fetch('http://my-cpanel-server.com:2087', {
 
 ### JavaScript
 
-```js
-var api = require('@cpanel/api');
-```
-
 #### Calling a WHM Api 1 function
```
Comment (laurencesimoncpanel) [inline README.md]: Api -> API

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@cpanel/api",
-  "version": "3.0.9",
+  "version": "3.0.10",
   "description": "cPanel Api JavaScript and TypeScript interface libraries. This library provides a set of classes for calling cPanel Whm API 1 and UAPI calls. The classes hide much of the complexity of these API's behind classes the abstract the underlying variances between these API systems. Users of this library can focus on what they want to accomplish rather the having to learn the various complexities of the underlying wire formats for each of the cPanel Products APIs",
```
Comment (laurencesimoncpanel) [inline package.json]: Api -> API
Whm -> WHM
API's -> APIs

File: src/interchange.ts [hunk 6]
```diff
@@ -44,7 +30,7 @@ export interface RequestInfo {
     /**
      * List of headers for the request
      */
-    headers: Header[];
+    headers: Headers;
 
     /**
      * Url use to make the request.
```
Comment (laurencesimoncpanel) [inline src/interchange.ts]: Url -> URL

File: src/request.ts [hunk 7]
```diff
@@ -299,6 +315,21 @@ export abstract class Request {
         return this;
     }
 
+    /**
+     * Add a custom http header to the request
```
Comment (laurencesimoncpanel) [inline src/request.ts:319]: http -> HTTP

File: src/uapi/request.spec.ts [hunk 8]
```diff
@@ -192,13 +199,79 @@ describe("UapiRequest", () => {
...
                     value: "application/json"
-                }],
+                }]),
                 url: "/execute/test/get_tests_by_label",
                 body: '{"label":"unit"}',
             });
         });
     });
+
+    describe("when calling with cpanel api token with token and user", () => {
```
Comment (laurencesimoncpanel) [inline src/uapi/request.spec.ts]: cpanel -> cPanel
api -> API

File: src/uapi/request.spec.ts [hunk 9]
```diff
@@ -192,13 +199,79 @@ describe("UapiRequest", () => {
...
+                    value: "cpanel user:fake"
+                }
+                ]),
+                url: "/execute/test/simple_call",
+                body: "",
+            });
+        });
+    });
+
+    describe("when calling with cpanel api token with combined user/token", () => {
```
Comment (laurencesimoncpanel) [inline src/uapi/request.spec.ts]: cpanel -> cPanel
api -> API

File: src/uapi/request.spec.ts [hunk 10]
```diff
@@ -192,13 +199,79 @@ describe("UapiRequest", () => {
...
+                    value: "cpanel user:fake"
+                }
+                ]),
+                url: "/execute/test/simple_call",
+                body: "",
+            });
+        });
+    });
+
+    describe("when calling with whm api token", () => {
```
Comment (laurencesimoncpanel) [inline src/uapi/request.spec.ts]: whm -> WHM
api -> API

File: src/uapi/request.ts [hunk 11]
```diff
@@ -68,9 +75,22 @@ import {
     WwwFormUrlArgumentEncoder
 } from "../utils/encoders";
 
-
 export class UapiRequest extends Request {
 
+    /**
+     * Add a custom http header to the request
```
Comment (laurencesimoncpanel) [inline src/uapi/request.ts]: http -> HTTP

File: src/utils/headers.ts [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+/**
+ * Http Header Abstraction
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Http -> HTTP

File: src/utils/headers.ts [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    /**
+     * Value of the header
+     */
+    value: string;
+}
+
+export type HeaderHash = { [index:string] : string };
+
+/**
+ * Http Headers Collection Abstraction
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Http -> HTTP

File: src/utils/headers.ts [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+ *
+ * The abstraction is an adaptor to transform the headers array into various
+ * formats for external http libraries.
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: http -> HTTP

File: src/utils/headers.ts [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+ */
+export class Headers {
+    private headers: Header[];
+
+    /**
+     * Create the adaptor.
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Not sure of the context, should adaptor be adapter?

File: src/utils/headers.ts [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+    }
+}
+
+export class CpanelApiTokenHeader extends CustomHeader {
+    constructor(token: string, user?: string) {
+        if (!token) {
+            throw new CpanelApiTokenInvalidError("You must pass a valid token to the constructor.");
+        }
+        if (!user && !/^.+[:]/.test(token)) {
+            throw new CpanelApiTokenInvalidError("You must pass a cPanel username associated with the cPanel Api token.");
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Api -> API

File: src/utils/headers.ts [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        if (!user && !/[:].+$/.test(token)) {
+            throw new CpanelApiTokenInvalidError("You must pass a valid cPanel Api token.");
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Api -> API

File: src/utils/headers.ts [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+    }
+}
+
+export class WhmApiTokenHeader extends CustomHeader {
+    constructor(token: string, user?: string) {
+        if (!token) {
+            throw new WhmApiTokenInvalidError("You must pass a valid token to the constructor.");
+        }
+        if (!user && !/^.+:/.test(token)) {
+            throw new WhmApiTokenInvalidError("You must pass a Whm username associated with the Whm Api token.");
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Whm -> WHM (both times)
Api -> API

File: src/utils/headers.ts [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        if (!user && !/:.+$/.test(token)) {
+            throw new WhmApiTokenInvalidError("You must pass a valid Whm Api token.");
```
Comment (laurencesimoncpanel) [inline src/utils/headers.ts]: Whm -> WHM
Api -> API

File: src/whmapi/request.spec.ts [hunk 20]
```diff
@@ -186,13 +193,79 @@ describe("WhmApiRequest: ", () => {
...
                     value: "application/json"
-                }],
+                }]),
                 url: "/json-api/api_method",
                 body: '{"api.version":1,"label":"unit"}',
             });
         });
     });
+
+    describe("when calling with whm api token with token and user", () => {
```
Comment (laurencesimoncpanel) [inline src/whmapi/request.spec.ts]: whm -> WHM
api -> API

File: src/whmapi/request.spec.ts [hunk 21]
```diff
@@ -186,13 +193,79 @@ describe("WhmApiRequest: ", () => {
...
+                    value: "whm user:fake"
+                }
+                ]),
+                url: "/json-api/simple_call",
+                body: "api.version=1",
+            });
+        });
+    });
+
+    describe("when calling with whm api token with combined user/token", () => {
```
Comment (laurencesimoncpanel) [inline src/whmapi/request.spec.ts]: whm -> WHM
api -> API

File: src/whmapi/request.spec.ts [hunk 22]
```diff
@@ -186,13 +193,79 @@ describe("WhmApiRequest: ", () => {
...
+                    value: "whm user:fake"
+                }
+                ]),
+                url: "/json-api/simple_call",
+                body: "api.version=1",
+            });
+        });
+    });
+
+    describe("when calling with cpanel api token", () => {
```
Comment (laurencesimoncpanel) [inline src/whmapi/request.spec.ts]: cpanel -> cPanel
api -> API

File: src/whmapi/request.ts [hunk 23]
```diff
@@ -93,6 +100,20 @@ export class WhmApiRequest extends Request {
      */
     public apiType: WhmApiType = WhmApiType.JsonApi;
 
+    /**
+     * Add a custom http header to the request
```
Comment (laurencesimoncpanel) [inline src/whmapi/request.ts]: http -> HTTP

**Review by Brianm0823 — APPROVED** [review]
Looks good to me...

File: CHANGELOG.md (hunk 1 above)
Comment (tomgreen98) [reply CHANGELOG.md]: Fixed.

File: README.md (hunk 2 above)
Comment (tomgreen98) [reply README.md]: Fixed.

File: README.md (hunk 3 above)
Comment (tomgreen98) [reply README.md]: Fixed.

File: README.md (hunk 4 above)
Comment (tomgreen98) [reply README.md]: Fixed.

File: package.json (hunk 5 above)
Comment (tomgreen98) [reply package.json]: Fixed.

File: src/interchange.ts (hunk 6 above)
Comment (tomgreen98) [reply src/interchange.ts]: Fixed.

File: src/uapi/request.spec.ts (hunk 8 above)
Comment (tomgreen98) [reply src/uapi/request.spec.ts]: Fixed.

File: src/uapi/request.spec.ts (hunk 10 above)
Comment (tomgreen98) [reply src/uapi/request.spec.ts]: Fixed.

File: src/uapi/request.ts (hunk 11 above)
Comment (tomgreen98) [reply src/uapi/request.ts]: Fixed.

File: src/utils/headers.ts (hunk 13 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed.

File: src/utils/headers.ts (hunk 14 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed

File: src/utils/headers.ts (hunk 15 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed.

File: src/utils/headers.ts (hunk 16 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed.

File: src/utils/headers.ts (hunk 18 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed.

File: src/utils/headers.ts (hunk 19 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed.

File: src/whmapi/request.spec.ts (hunk 20 above)
Comment (tomgreen98) [reply src/whmapi/request.spec.ts]: Fixed.

File: src/whmapi/request.spec.ts (hunk 21 above)
Comment (tomgreen98) [reply src/whmapi/request.spec.ts]: Fixed.

File: src/whmapi/request.ts (hunk 23 above)
Comment (tomgreen98) [reply src/whmapi/request.ts]: Fixed.

File: src/whmapi/request.spec.ts (hunk 22 above)
Comment (tomgreen98) [reply src/whmapi/request.spec.ts]: Fixed.

File: src/utils/headers.ts (hunk 12 above)
Comment (tomgreen98) [reply src/utils/headers.ts]: Fixed.

File: src/uapi/request.spec.ts (hunk 9 above)
Comment (tomgreen98) [reply src/uapi/request.spec.ts]: Fixed.

File: src/request.ts (hunk 7 above)
Comment (tomgreen98) [reply src/request.ts:319]: Fixed.

File: README.md [hunk 24]
```diff
@@ -1,6 +1,6 @@
 # @cpanel/api Libraries
 
-cPanel Api JavaScript and TypeScript interface libraries.
+cPanel API JavaScript and TypeScript interface libraries.
 
 This library provides a set of classes for calling cPanel Whm API 1 and UAPI calls. The classes hide much of the complexity of these API's behind classes the abstract the underlying variances between these API systems. Users of this library can focus on what they want to accomplish rather the having to learn the various complexities of the underlying wire formats for each of the cPanel Products APIs
```
Comment (laurencesimoncpanel) [inline README.md]: Whm -> WHM

Comment (laurencesimoncpanel) [inline README.md]: Products -> product's ?

**Review by laurencesimoncpanel — COMMENTED** [review]
Muuuuuuch cleaner ty... just two observations I must have missed and good to go.
ty
-ls

File: README.md (hunk 24 above)
Comment (tomgreen98) [reply README.md]: Fixed.

Comment (tomgreen98) [reply README.md]: Fixed.

**Review by laurencesimoncpanel — APPROVED** [review]

**Pull request merged**